File: lib/ruby/cosmos-llm-ruby-tool-preset/lib/cosmos/llm/tool/preset/grep.rb
# frozen_string_literal: true

require 'cosmos/llm/tool'

module Cosmos
  module Llm
    module Tool
      module Preset
        # Grep (search) preset tool (virtual filesystem)
        #
        # Provides functionality to search for patterns in files within the
        # virtual filesystem using regular expressions.
        #
        # @example Using the grep tool
        #   tool = Cosmos::Llm::Tool::Preset.grep(filesystem)
        #   result = tool.call(pattern: 'TODO', file_pattern: '*.rb')
        #
        # @param filesystem [Cosmos::Llm::VirtualFilesystem::Filesystem] The virtual filesystem to search
        # @return [Cosmos::Llm::Tool::Definition] A grep search tool
        def self.grep(filesystem)
          Cosmos::Llm::Tool.define(:grep, register: false) do
            description 'Search for patterns in files using regular expressions within the virtual filesystem'

            parameter :pattern,
                      type: :string,
                      required: true,
                      description: 'The regex pattern to search for in file contents'

            parameter :file_pattern,
                      type: :string,
                      required: false,
                      description: 'File pattern to include in search (e.g., "*.rb", "*.{js,ts}")'

            execute do |params|
              search_pattern = params[:pattern]
              file_pattern = params[:file_pattern]

              begin
                regex = Regexp.new(search_pattern)

                # Get all files
                all_files = filesystem.all_files

                # Filter by file pattern if provided
                files_to_search = if file_pattern
                                    # Convert glob pattern to regex (simple conversion for search)
                                    regex_pattern = file_pattern.dup
                                    regex_pattern = regex_pattern.gsub(/([.+^$|()\[\]])/) do
                                      "\\#{Regexp.last_match(1)}"
                                    end
                                    regex_pattern = regex_pattern.gsub('*', '.*')
                                    regex_pattern = regex_pattern.gsub('?', '.')
                                    # Handle {a,b} alternation
                                    regex_pattern = regex_pattern.gsub(/\{([^}]+)\}/) do
                                      options = Regexp.last_match(1).split(',')
                                      "(#{options.join('|')})"
                                    end
                                    file_regex = Regexp.new(regex_pattern)
                                    all_files.select { |file_info| file_regex.match?(file_info[:path]) }
                                  else
                                    all_files
                                  end

                # Search through files
                matches = []
                files_searched = 0
                files_with_matches = 0

                files_to_search.each do |file_info|
                  file = file_info[:file]
                  path = file_info[:path]
                  content = file.content

                  # Skip binary files or files without content
                  next unless content
                  next if content.include?("\x00")

                  files_searched += 1
                  lines = content.lines
                  file_has_matches = false

                  lines.each_with_index do |line, index|
                    line.scan(regex) do
                      file_has_matches = true
                      matches << {
                        file: path,
                        line_number: index + 1,
                        content: line.chomp,
                        match: Regexp.last_match[0]
                      }
                    end
                  end

                  files_with_matches += 1 if file_has_matches
                end

                {
                  success: true,
                  pattern: search_pattern,
                  file_pattern: file_pattern,
                  matches: matches,
                  match_count: matches.length,
                  files_searched: files_searched,
                  files_with_matches: files_with_matches
                }
              rescue RegexpError => e
                {
                  success: false,
                  error: "Invalid regex pattern: #{e.message}",
                  pattern: search_pattern
                }
              rescue StandardError => e
                {
                  success: false,
                  error: e.message,
                  pattern: search_pattern
                }
              end
            end
          end
        end
      end
    end
  end
end

# Copyright (c) 2025 Durable Programming, LLC. All rights reserved.

File: lib/rust/cosmos-llm-rust-tool/src/preset/grep.rs
use std::sync::Arc;

use cosmos_llm_virtual_filesystem::Filesystem;
use regex::Regex;
use serde_json::json;

use crate::{ParameterType, ToolDefinition};

/// Creates a grep preset tool bound to `filesystem`.
///
/// Searches for a regex pattern across all (or a filtered subset of) files in
/// the virtual filesystem. Binary files (containing null bytes) are skipped.
///
/// # Examples
///
/// ```rust
/// use cosmos_llm_virtual_filesystem::Filesystem;
/// use cosmos_llm_tool::preset::grep_tool;
/// use serde_json::json;
/// use std::sync::Arc;
///
/// let fs = Arc::new(Filesystem::build("/", |root| {
///     root.file("a.rs", Some("// TODO: fix this\nfn main(){}\n"), Default::default()).unwrap();
/// }));
///
/// let tool = grep_tool(fs);
/// let r = tool.call(&json!({ "pattern": "TODO" })).unwrap();
/// assert_eq!(r["match_count"], json!(1));
/// ```
pub fn grep_tool(filesystem: Arc<Filesystem>) -> ToolDefinition {
    ToolDefinition::new("grep")
        .description(
            "Search for regex patterns in files within the virtual filesystem",
        )
        .param(
            "pattern",
            ParameterType::String,
            true,
            "Regex pattern to search for",
        )
        .param(
            "file_pattern",
            ParameterType::String,
            false,
            "Glob pattern to restrict which files are searched (e.g. '*.rb')",
        )
        .handler(move |params| {
            let search_pattern = params["pattern"]
                .as_str()
                .ok_or("pattern must be a string")?
                .to_string();
            let file_pattern = params["file_pattern"].as_str().map(|s| s.to_string());

            let regex = Regex::new(&search_pattern)
                .map_err(|e| format!("invalid regex pattern: {e}"))?;

            let all = filesystem.all_files("");

            let files_to_search: Vec<_> = if let Some(ref fpat) = file_pattern {
                let freg = file_glob_regex(fpat)
                    .map_err(|e| format!("invalid file pattern: {e}"))?;
                let match_basename = !fpat.contains('/');
                all.into_iter()
                    .filter(|e| {
                        let candidate = if match_basename {
                            e.path.rsplit('/').next().unwrap_or(&e.path)
                        } else {
                            e.path.as_str()
                        };
                        freg.is_match(candidate)
                    })
                    .collect()
            } else {
                all
            };

            let mut matches: Vec<serde_json::Value> = Vec::new();
            let mut files_searched = 0usize;
            let mut files_with_matches = 0usize;

            for entry in &files_to_search {
                let content = match &entry.file.content {
                    Some(c) if !c.contains('\x00') => c,
                    _ => continue,
                };

                files_searched += 1;
                let mut file_matched = false;

                for (idx, line) in content.lines().enumerate() {
                    for m in regex.find_iter(line) {
                        file_matched = true;
                        matches.push(json!({
                            "file": entry.path,
                            "line_number": idx + 1,
                            "content": line,
                            "match": m.as_str(),
                        }));
                    }
                }

                if file_matched {
                    files_with_matches += 1;
                }
            }

            Ok(json!({
                "success": true,
                "pattern": search_pattern,
                "file_pattern": file_pattern,
                "matches": matches,
                "match_count": matches.len(),
                "files_searched": files_searched,
                "files_with_matches": files_with_matches,
            }))
        })
}

fn file_glob_regex(pat: &str) -> Result<Regex, regex::Error> {
    let mut out = String::from("^");
    let mut chars = pat.chars().peekable();
    while let Some(c) = chars.next() {
        match c {
            '*' if chars.peek() == Some(&'*') => {
                chars.next();
                if chars.peek() == Some(&'/') {
                    chars.next();
                }
                out.push_str(".*");
            }
            '*' => out.push_str("[^/]*"),
            '.' | '+' | '^' | '$' | '(' | ')' | '[' | ']' | '{' | '}' | '|' | '\\' => {
                out.push('\\');
                out.push(c);
            }
            _ => out.push(c),
        }
    }
    out.push('$');
    Regex::new(&out)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn make_fs() -> Arc<Filesystem> {
        Arc::new(Filesystem::build("/", |root| {
            root.file(
                "a.rs",
                Some("// TODO: fix\nfn main(){}\n"),
                Default::default(),
            )
            .unwrap();
            root.file("b.txt", Some("nothing here\n"), Default::default())
                .unwrap();
        }))
    }

    #[test]
    fn finds_match() {
        let t = grep_tool(make_fs());
        let r = t.call(&json!({ "pattern": "TODO" })).unwrap();
        assert_eq!(r["match_count"], json!(1));
        assert_eq!(r["files_with_matches"], json!(1));
    }

    #[test]
    fn file_pattern_filters() {
        let t = grep_tool(make_fs());
        let r = t
            .call(&json!({ "pattern": "nothing", "file_pattern": "*.rs" }))
            .unwrap();
        assert_eq!(r["match_count"], json!(0));
    }

    #[test]
    fn invalid_regex_fails() {
        let t = grep_tool(make_fs());
        let r = t.call(&json!({ "pattern": "[invalid" }));
        assert!(r.is_err() || r.unwrap()["success"] == json!(false));
    }
}
